File: apps/web/lib/format.ts
/**
 * Formatage d'affichage — AUCUN calcul monétaire ici (CLAUDE.md §1.1 : jamais d'arithmétique
 * côté client). On ne fait que parser des chaînes décimales de l'API pour les présenter :
 * `1 250,00 MAD`, chiffres tabulaires gérés par la classe CSS `.tnum`.
 */
import type { Locale } from "./i18n";

/** "1250.5" | "1250" | "1250.00" → "1 250,00" (séparateur fin insécable). */
export function formatMontant(value: string | null | undefined): string {
  if (value === null || value === undefined || value === "") return "—";
  const negative = value.startsWith("-");
  const abs = negative ? value.slice(1) : value;
  const [intPartRaw = "0", decPart = ""] = abs.split(".");
  const intPart = intPartRaw.replace(/^0+(?=\d)/, "");
  const grouped = intPart.replace(/\B(?=(\d{3})+(?!\d))/g, "\u202f");
  const dec = (decPart + "00").slice(0, 2);
  return `${negative ? "−" : ""}${grouped},${dec}`;
}

export function formatMAD(value: string | null | undefined, locale: Locale): string {
  if (value === null || value === undefined || value === "") return "—";
  const montant = formatMontant(value);
  return locale === "ar" ? `${montant} د.م.` : `${montant} MAD`;
}

/** Nombre entier "1000" → "1 000" (tantièmes, compteurs). */
export function formatEntier(value: string | number | null | undefined): string {
  if (value === null || value === undefined || value === "") return "—";
  const s = String(value).split(".")[0] ?? "0";
  return s.replace(/\B(?=(\d{3})+(?!\d))/g, "\u202f");
}

/** Ratio "0.5" → "50 %". */
export function formatPourcent(ratio: string | number | null | undefined): string {
  if (ratio === null || ratio === undefined || ratio === "") return "—";
  const n = typeof ratio === "number" ? ratio : Number.parseFloat(ratio);
  if (Number.isNaN(n)) return "—";
  return `${Math.round(n * 1000) / 10}\u202f%`;
}

const INTL_LOCALE: Record<Locale, string> = { fr: "fr-MA", ar: "ar-MA" };

export function formatDate(iso: string | null | undefined, locale: Locale): string {
  if (!iso) return "—";
  const d = new Date(iso);
  if (Number.isNaN(d.getTime())) return "—";
  return new Intl.DateTimeFormat(INTL_LOCALE[locale], {
    day: "numeric",
    month: "long",
    year: "numeric",
    numberingSystem: "latn",
  }).format(d);
}

export function formatDateCourte(iso: string | null | undefined, locale: Locale): string {
  if (!iso) return "—";
  const d = new Date(iso);
  if (Number.isNaN(d.getTime())) return "—";
  return new Intl.DateTimeFormat(INTL_LOCALE[locale], {
    day: "2-digit",
    month: "2-digit",
    year: "numeric",
    numberingSystem: "latn",
  }).format(d);
}

export function formatDateHeure(iso: string | null | undefined, locale: Locale): string {
  if (!iso) return "—";
  const d = new Date(iso);
  if (Number.isNaN(d.getTime())) return "—";
  return new Intl.DateTimeFormat(INTL_LOCALE[locale], {
    day: "numeric",
    month: "short",
    year: "numeric",
    hour: "2-digit",
    minute: "2-digit",
    numberingSystem: "latn",
  }).format(d);
}

export function formatHeure(iso: string | null | undefined, locale: Locale): string {
  if (!iso) return "—";
  const d = new Date(iso);
  if (Number.isNaN(d.getTime())) return "—";
  return new Intl.DateTimeFormat(INTL_LOCALE[locale], {
    hour: "2-digit",
    minute: "2-digit",
    numberingSystem: "latn",
  }).format(d);
}

/** Période "2026-01" → "janvier 2026". */
export function formatPeriode(periode: string, locale: Locale): string {
  const [y, m] = periode.split("-");
  if (!y || !m) return periode;
  const d = new Date(Date.UTC(Number(y), Number(m) - 1, 1));
  return new Intl.DateTimeFormat(INTL_LOCALE[locale], {
    month: "long",
    year: "numeric",
    numberingSystem: "latn",
  }).format(d);
}

/** Téléphone +212612345678 → +212 6 12 34 56 78 (affichage). */
export function formatTelephone(tel: string | null | undefined): string {
  if (!tel) return "—";
  const m = tel.match(/^\+?212(\d)(\d{2})(\d{2})(\d{2})(\d{2})$/);
  if (m) return `+212 ${m[1]} ${m[2]} ${m[3]} ${m[4]} ${m[5]}`;
  return tel;
}

export function nomComplet(u: { nom?: string | null; prenom?: string | null } | null | undefined): string | null {
  if (!u) return null;
  const s = [u.prenom, u.nom].filter(Boolean).join(" ");
  return s || null;
}

/** Nombre de jours (calendaires) entre maintenant et une date ISO — négatif si passée. */
export function joursRestants(iso: string): number {
  const cible = new Date(iso);
  const debutCible = Date.UTC(cible.getUTCFullYear(), cible.getUTCMonth(), cible.getUTCDate());
  const now = new Date();
  const debutNow = Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate());
  return Math.round((debutCible - debutNow) / 86_400_000);
}

File: apps/mobile/lib/core/format/format.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';

/// Formatage d'affichage — AUCUN calcul monétaire ici (CLAUDE.md §1.1). Port de
/// apps/web/lib/format.ts pour un rendu identique sur les deux clients.

const String _nbsp = ' ';

/// "1250.5" | "1250" | "1250.00" → "1 250,00".
String formatMontant(String? value) {
  if (value == null || value.isEmpty) return '—';
  final negative = value.startsWith('-');
  final abs = negative ? value.substring(1) : value;
  final parts = abs.split('.');
  final intRaw = parts.isNotEmpty && parts[0].isNotEmpty ? parts[0] : '0';
  final decPart = parts.length > 1 ? parts[1] : '';
  final intPart = intRaw.replaceFirst(RegExp(r'^0+(?=\d)'), '');
  final grouped = _group(intPart);
  final dec = ('${decPart}00').substring(0, 2);
  return '${negative ? '−' : ''}$grouped,$dec';
}

String _group(String digits) {
  final buf = StringBuffer();
  for (int i = 0; i < digits.length; i++) {
    final left = digits.length - i;
    buf.write(digits[i]);
    if (left > 1 && left % 3 == 1) buf.write(_nbsp);
  }
  return buf.toString();
}

String formatMAD(String? value, Locale locale) {
  if (value == null || value.isEmpty) return '—';
  final m = formatMontant(value);
  return locale.languageCode == 'ar' ? '$m د.م.' : '$m MAD';
}

/// Entier "1000" → "1 000" (tantièmes, compteurs).
String formatEntier(Object? value) {
  if (value == null || value.toString().isEmpty) return '—';
  final s = value.toString().split('.').first;
  return _group(s);
}

/// Ratio 0.5 → "50 %".
String formatPourcent(double? r) {
  if (r == null) return '—';
  final v = (r * 1000).round() / 10;
  final s = v == v.roundToDouble() ? v.toInt().toString() : v.toString();
  return '$s$_nbsp%';
}

String _intlLocale(Locale l) => l.languageCode == 'ar' ? 'ar' : 'fr';

/// intl rend les chiffres arabes-indiens en `ar` ; le produit garde les chiffres latins
/// (parité web `numberingSystem: latn`).
String latn(String s) {
  const from = '٠١٢٣٤٥٦٧٨٩';
  final b = StringBuffer();
  for (final r in s.runes) {
    final ch = String.fromCharCode(r);
    final i = from.indexOf(ch);
    b.write(i >= 0 ? i.toString() : ch);
  }
  return b.toString();
}

DateTime? _parse(String? iso) {
  if (iso == null || iso.isEmpty) return null;
  return DateTime.tryParse(iso)?.toLocal();
}

String formatDate(String? iso, Locale locale) {
  final d = _parse(iso);
  if (d == null) return '—';
  return latn(DateFormat.yMMMMd(_intlLocale(locale)).format(d));
}

String formatDateCourte(String? iso, Locale locale) {
  final d = _parse(iso);
  if (d == null) return '—';
  return latn(DateFormat('dd/MM/yyyy', _intlLocale(locale)).format(d));
}

String formatDateHeure(String? iso, Locale locale) {
  final d = _parse(iso);
  if (d == null) return '—';
  final l = _intlLocale(locale);
  return latn('${DateFormat.MMMd(l).format(d)} ${DateFormat.y(l).format(d)} · ${DateFormat.Hm(l).format(d)}');
}

String formatHeure(String? iso, Locale locale) {
  final d = _parse(iso);
  if (d == null) return '—';
  return latn(DateFormat.Hm(_intlLocale(locale)).format(d));
}

/// Période "2026-01" → "janvier 2026".
String formatPeriode(String periode, Locale locale) {
  final p = periode.split('-');
  if (p.length < 2) return periode;
  final y = int.tryParse(p[0]);
  final m = int.tryParse(p[1]);
  if (y == null || m == null) return periode;
  return latn(DateFormat.yMMMM(_intlLocale(locale)).format(DateTime(y, m, 1)));
}

/// Jour de semaine long + date + heure ("vendredi 26 septembre 2026 · 18:30").
String formatDateLongue(String? iso, Locale locale) {
  final d = _parse(iso);
  if (d == null) return '—';
  final l = _intlLocale(locale);
  return latn('${DateFormat.yMMMMEEEEd(l).format(d)} · ${DateFormat.Hm(l).format(d)}');
}

/// Téléphone +212612345678 → +212 6 12 34 56 78.
String formatTelephone(String? tel) {
  if (tel == null || tel.isEmpty) return '—';
  final m = RegExp(r'^\+?212(\d)(\d{2})(\d{2})(\d{2})(\d{2})$').firstMatch(tel);
  if (m != null) return '+212 ${m[1]} ${m[2]} ${m[3]} ${m[4]} ${m[5]}';
  return tel;
}

String? nomComplet(String? prenom, String? nom) {
  final s = [prenom, nom].where((x) => x != null && x.isNotEmpty).join(' ');
  return s.isEmpty ? null : s;
}

/// Jours calendaires entre aujourd'hui et une date ISO — négatif si passée.
int joursRestants(String iso) {
  final cible = DateTime.parse(iso).toLocal();
  final now = DateTime.now();
  final a = DateTime(cible.year, cible.month, cible.day);
  final b = DateTime(now.year, now.month, now.day);
  return a.difference(b).inDays;
}

bool estAujourdhui(String iso) {
  final d = DateTime.tryParse(iso)?.toLocal();
  if (d == null) return false;
  final n = DateTime.now();
  return d.year == n.year && d.month == n.month && d.day == n.day;
}

/// Téléphone marocain saisi librement (06…, +2126…, 002126…) → +2126XXXXXXXX.
String? normaliserTelephone(String brut) {
  var t = brut.replaceAll(RegExp(r'[^\d+]'), '');
  if (t.startsWith('00212')) {
    t = '+212${t.substring(5)}';
  } else if (t.startsWith('212')) {
    t = '+$t';
  } else if (t.startsWith('0') && t.length == 10) {
    t = '+212${t.substring(1)}';
  }
  if (RegExp(r'^\+2126\d{8}$').hasMatch(t) || RegExp(r'^\+2127\d{8}$').hasMatch(t)) return t;
  return null;
}
